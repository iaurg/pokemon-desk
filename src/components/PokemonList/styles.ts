import styled from 'styled-components'

export const Wrapper = styled.main`
  margin: 2rem auto;
`
export const Loading = styled.div``

export const Error = styled.div``

export const ListContent = styled.div`
  display: flex;
  flex-wrap: wrap;
`
