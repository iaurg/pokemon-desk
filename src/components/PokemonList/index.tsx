import PokemonCard from 'components/PokemonCard'
import { useQuery } from 'react-query'
import { getPokemon, Pokemon as PokemonType } from 'services/getPokemon'
import { getPokemons } from 'services/getPokemons'
import * as S from './styles'

function Pokemon({ name }: { name: string }) {
  const { data, isLoading, isError } = useQuery(
    ['pokemon', name],
    () => getPokemon(name),
    {
      staleTime: 1000 * 60 * 1 // 1 minute
    }
  )

  if (isLoading) {
    return <S.LoadingCard>Loading</S.LoadingCard>
  }

  if (isError) {
    return <S.LoadingCard>Error</S.LoadingCard>
  }

  return (
    <PokemonCard
      name={data?.name || ''}
      types={data?.types || []}
      image={data?.sprites.other['official-artwork'].front_default || ''}
    />
  )
}

function PokemonList({ search = '' }) {
  const pokemons = useQuery(['pokemons'], () => getPokemons(), {
    staleTime: 1000 * 60 * 1 // 1 minutes
  })

  if (pokemons.isLoading) {
    return <S.Loading>Loading...</S.Loading>
  }

  if (pokemons.isError) {
    return <S.Error>Error when fetching pokemons</S.Error>
  }

  if (search && pokemons?.data) {
    const filteredPokemons = pokemons.data.results.filter(
      (pokemon: Pick<PokemonType, 'name'>) =>
        pokemon.name.toLowerCase().includes(search.toLowerCase())
    )

    return (
      <S.Wrapper>
        <h1>PokemonList</h1>
        <S.ListContent>
          {filteredPokemons.map((pokemon) => (
            <Pokemon key={pokemon.name} name={pokemon.name} />
          ))}
          {filteredPokemons.length === 0 && <>Nothing found... Try again</>}
        </S.ListContent>
      </S.Wrapper>
    )
  }

  return (
    <S.Wrapper>
      <h1>PokemonList</h1>
      <S.ListContent>
        {pokemons.data?.results.map((pokemon) => (
          <Pokemon key={pokemon.name} name={pokemon.name} />
        ))}
      </S.ListContent>
    </S.Wrapper>
  )
}

export default PokemonList
