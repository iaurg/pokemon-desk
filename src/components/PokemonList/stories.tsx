import { Story, Meta } from '@storybook/react'
import PokemonList from '.'

export default {
  title: 'PokemonList',
  component: PokemonList
} as Meta

export const Default: Story = () => <PokemonList />
