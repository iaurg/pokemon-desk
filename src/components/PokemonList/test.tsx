import { render, screen } from '@testing-library/react'

import PokemonList from '.'

describe('<PokemonList />', () => {
  it('should render the heading', () => {
    const { container } = render(<PokemonList />)

    expect(
      screen.getByRole('heading', { name: /PokemonList/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
