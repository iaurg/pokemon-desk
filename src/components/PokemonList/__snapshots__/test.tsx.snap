// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonList /> should render the heading 1`] = `
<main
  class=""
>
  <h1>
    PokemonList
  </h1>
</main>
`
